src/Lean/Parser/Basic.lean

```diff
@@ -856,25 +856,30 @@ where
       let i    := c.next i
       let curr := c.get i
       if curr.isDigit then
-        takeDigitsFn (fun c => c.isDigit) "decimal number" false c (s.setPos i)
+        (takeDigitsFn (fun c => c.isDigit) "decimal number" false c (s.setPos i), true)
       else
-        s.setPos i
+        (s.setPos i, true)
     else
-      s
+      (s, false)
 
-  parseOptExp s :=
-    let i     := s.pos
-    let curr  := c.get i
-    if curr == 'e' || curr == 'E' then
-      let i    := c.next i
-      let i    := if c.get i == '-' || c.get i == '+' then c.next i else i
-      let curr := c.get i
-      if curr.isDigit then
-        takeDigitsFn (fun c => c.isDigit) "decimal number" false c (s.setPos i)
-      else
-        s.mkUnexpectedError "missing exponent digits in scientific literal"
-    else
+  parseOptExp s hasDot :=
+    if h : c.atEnd s.pos then
       s
+    else
+      let i     := s.pos
+      let curr  := c.get' i h
+      if curr == 'e' || curr == 'E' then
+        let i    := c.next i
+        let i    := if c.get i == '-' || c.get i == '+' then c.next i else i
+        let curr := c.get i
+        if curr.isDigit then
+          takeDigitsFn (fun c => c.isDigit) "decimal number" false c (s.setPos i)
+        else
+          s.mkUnexpectedError "missing exponent digits in scientific literal"
+      else if hasDot && ! c.atEnd s.pos && isIdFirst curr then
+          (s.setPos startPos).mkUnexpectedError s!"unexpected identifier after decimal point; consider parenthesizing the number"
```

Comment:
The `s.setPos startPos` here ensures that the tokenizer is not productive: if you just put `s` it will parse as <scientificLit> <ERROR: unexpected identifier after decimal point>, which casuses _exactly one_ problem in the test suite: this term goes from parsing correctly to generating that new error message:

```
{ foo with x[1].1.fst := 1 }
```

I intuited that this was probably a backtracking-parsing issue; it's fixed by setting the position so that the error is placed at the start of the numeral.

I'm holding off on merging this PR until I can better understand exactly what happened here and create a better, hopefully more minimal, testing breadcrumb.